edb/server/protocol/auth_ext/http.py

```diff
@@ -70,995 +72,49 @@ async def handle_request(
-                        response.custom_headers[
-                            "Location"
-                        ] = f"{redirect_to}?{urllib.parse.urlencode(params)}"
-                        response.status = http.HTTPStatus.FOUND
-                        return
-
-                    if code is None:
-                        raise errors.InvalidData(
-                            "Provider did not include the 'code' parameter in "
-                            "callback"
-                        )
-
-                    try:
-                        claims = self._verify_and_extract_claims(state)
-                        provider_name = cast(str, claims["provider"])
-                        redirect_to = cast(str, claims["redirect_to"])
-                        redirect_to_on_signup = cast(
-                            Optional[str], claims.get("redirect_to_on_signup")
-                        )
-                        challenge = cast(str, claims["challenge"])
-                    except Exception:
-                        raise errors.InvalidData("Invalid state token")
-                    oauth_client = oauth.Client(
-                        db=self.db,
-                        provider_name=provider_name,
-                        base_url=test_url,
-                    )
-                    (
-                        identity,
-                        new_identity,
-                        auth_token,
-                        refresh_token,
-                    ) = await oauth_client.handle_callback(
-                        code, self._get_callback_url()
-                    )
-                    pkce_code = await pkce.link_identity_challenge(
-                        self.db, identity.id, challenge
-                    )
-                    if auth_token or refresh_token:
-                        await pkce.add_provider_tokens(
-                            self.db,
-                            id=pkce_code,
-                            auth_token=auth_token,
-                            refresh_token=refresh_token,
-                        )
-                    parsed_url = urllib.parse.urlparse(
-                        (redirect_to_on_signup or redirect_to)
-                        if new_identity else redirect_to
-                    )
-                    query_params = urllib.parse.parse_qs(parsed_url.query)
-                    query_params["code"] = [pkce_code]
-                    new_query = urllib.parse.urlencode(query_params, doseq=True)
-                    new_url = parsed_url._replace(query=new_query).geturl()
-
-                    session_token = self._make_session_token(identity.id)
-                    response.status = http.HTTPStatus.FOUND
-                    response.custom_headers["Location"] = new_url
-                    _set_cookie(response, "edgedb-session", session_token)
-
+                    return await self.handle_callback(*handler_args)
```

Comment:
I also considered a dictionary of tuples to handlers, but this seemed a _bit_ more flexible just in case we needed to add any logic in the dispatch in the future. I'm very open to feedback here!

Reply:
As we're here, perhaps `edb.common.value_dispatch` would be nice too?

Reply:
Hmm.. Looking at the implementation and I'm not sure that meets the usage here: we're dispatch based on (for now exact) matches of a list of strings rather than the value of an argument. Maybe I'm misunderstanding how it'd be used?

Reply:
Not tested, but I'm thinking about something like this:

```python
    async def handle_request(
        self, request: Any, response: Any, args: list[str]
    ):
        ...
        try:
            return await self.handle(args, request, response)
        except errors.NotFound as ex:
            ...

    @value_dispatch.value_dispatch
    async def handle(self, args, request: Any, response: Any):
        ...

    @handle.register(("authorize",))
    async def handle_authorize(self, _args, request: Any, response: Any):
        ...
```

Reply:
You're right, `value_dispatch` doesn't support pattern bindings yet, but it shouldn't be hard to add a `pattern_dispatch` or something if needed. Before that, we could mix using `match` in a way like:

```python
    @value_dispatch.value_dispatch
    async def handle(self, args, request: Any, response: Any):
        match args:
            case('ui', '_static', filename):
                ...

            case _:
                raise errors.NotFound("Unknown auth endpoint")

    @handle.register(("authorize",))
    async def handle_authorize(self, _args, request: Any, response: Any):
        ...
```

Reply:
Oh that's so clever! I love it. Will do.

Reply:
@fantix The one thing that isn't yet obvious to me is what the `handle` function looks like so that it has the right fallback order. I'll play around with it and if I can't figure it out can I merge what we have here and revisit later?

Reply:
Alright, looks like it's a little bit of a larger refactor since this would require moving from a class to flat functions with a context struct, so I've spun that work off into a separate branch so this can land without being blocked.